### PR TITLE
Unsaved changes warning

### PR DESCRIPTION
**Pull request by fakob**
* Added an unsaved changes warning when
  * closing or reloading a tab/window -> shows standard close/reload tab browser dialog (no control over design or text)
  * loading another graph -> shows the browsers default confirmation dialog (no control over design)
    * for this case we can add a react modal dialog in the future to make it prettier

Screenshot 2023-02-03 at 20 20 29
Screenshot 2023-02-02 at 22 12 20
Screenshot 2023-02-03 at 20 16 05

### CODE REVIEW

File: src/PPStorage.tsx [hunk 1]
```diff
@@ -352,35 +379,31 @@ export default class PPStorage {
...
-          >
-            Dismiss
-          </Button>
-        </>
-      ),
-    });
+    if (checkForUnsavedChanges()) {
+      const nameOfFileToClone = remoteGraphsRef.current[id];
+      const fileData = await this.getRemoteGraph(nameOfFileToClone);
+      console.log(fileData);
```
Comment (magnificus) [inline src/PPStorage.tsx:385]: can remove log

File: src/classes/GraphClass.ts [hunk 2]
```diff
@@ -335,6 +336,11 @@ export default class PPGraph {
     }
   }
 
+  onBeforeUnload(event: BeforeUnloadEvent): string {
```
Comment (magnificus) [inline src/classes/GraphClass.ts:339]: not very clear what this function does, can it be made more explicit?

File: src/classes/GraphClass.ts [hunk 3]
```diff
@@ -48,6 +48,7 @@ export default class PPGraph {
 
   tempConnection: PIXI.Graphics;
   selection: PPSelection;
+  private unsavedChanges: boolean;
```
Comment (magnificus) [inline src/classes/GraphClass.ts:51]: could this instead be stored inside actionHandler? graphHasUnsavedChanges? cleaner to not expose implicit property this to the graph

File: src/utils/actionHandler.tsx [hunk 4]
```diff
@@ -1,46 +1,50 @@
...
+    action: Action,
+    undo: Action,
+    doPerformAction = true
+  ) {
+    if (doPerformAction) {
       await action();
     }
-    this.undoList.push({action:action, undo:undo});
+    this.undoList.push({ action: action, undo: undo });
+    PPGraph.currentGraph.setUnsavedChange(true);
```
Comment (magnificus) [inline src/utils/actionHandler.tsx:27]: lets keep it inside actionhandler instead if possible

**Review by magnificus — CHANGES_REQUESTED** [review]
Looks good and useful, but would strongly suggest putting the hasChanges variable inside action handler instead, to not let the graph know anything about how it is saved and why etc.

File: src/classes/GraphClass.ts (hunk 2 above)
Comment (fakob) [reply src/classes/GraphClass.ts]: added a comment

File: src/PPStorage.tsx (hunk 1 above)
Comment (fakob) [reply src/PPStorage.tsx]: done

File: src/classes/GraphClass.ts (hunk 3 above)
Comment (fakob) [reply src/classes/GraphClass.ts:51]: moved it to the actionhandler

File: src/PPStorage.tsx [hunk 5]
```diff
@@ -178,41 +200,44 @@ export default class PPStorage {
...
+      try {
+        PPGraph.currentGraph.configure(fileData);
 
-      // unset loadedGraphId
-      await this.db.settings.put({
-        name: 'loadedGraphId',
-        value: undefined,
-      });
+        // unset loadedGraphId
+        await this.db.settings.put({
```
Comment (magnificus) [inline src/PPStorage.tsx:209]: we should have a function "setLoadedGraphID" and "getLoadedGraphID" on PPStorage that does this, so we can hide the name in DB

**Review by magnificus — APPROVED** [review]

File: src/PPStorage.tsx (hunk 5 above)
Comment (fakob) [reply src/PPStorage.tsx]: done

**Pull request merged**